tests/cloudpickle_test.py

```diff
@@ -2130,6 +2137,38 @@ def test_pickle_importable_typevar(self):
         from typing import AnyStr
         assert AnyStr is pickle_depickle(AnyStr, protocol=self.protocol)
 
+    @unittest.skipIf(sys.version_info < (3, 7),
+                     "Pickling generics not supported below py37")
+    def test_generic_type(self):
+        for obj in _generic_objects_to_test():
+            pickle_depickle(obj, protocol=self.protocol)
+
+    @unittest.skipIf(sys.version_info < (3, 7),
+                     "Pickling type hints not supported below py37")
+    def test_locally_defined_class_with_type_hints(self):
+        with subprocess_worker(protocol=self.protocol) as worker:
+            for type_ in _generic_objects_to_test():
+                # The type annotation syntax causes a SyntaxError on Python 3.5
+                code = textwrap.dedent("""\
+                class MyClass:
+                    attribute: type_
```

Comment:
To make things clear: pickling `Generic` constructs works fine on master right now. The only thing this PR introduces is supporting the pickling of dynamic classes defined using the
```python
class A(typing.Generic[T]):
   ...
```
semantic. the type hint used in `attribute` as well as in `method` is correctly pickled (for `Python 3.7+`) in `cloudpickle` master as of now. I think adding such type hints in this tests generates more confusion that it helps.

Reply:
> To make things clear: pickling Generic constructs works fine on master right now.

Maybe, but it was not properly tested. So better take the opportunity to increase the test coverage in general, not just the specific fix from this PR.